feat: add C implementation for `stats/base/dists/levy/median`

vivekmaurya001 opened the pull request:
Resolves #3727.

## Description

- adds C implementation for @stdlib/stats/base/dists/levy/median along with relevant tests, benchmarks and examples

This pull request:

- Progresses: https://github.com/stdlib-js/stdlib/issues/3727
## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

 - [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)


* * *

@stdlib-js/reviewers 

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

vivekmaurya001 committed:
feat(add c implementation): add stats/base/dists/levy/median

vivekmaurya001 committed:
Merge branch 'levy/median' of https://github.com/vivekmaurya001/stdlib into levy/median

vivekmaurya001 committed:
Merge branch 'stdlib-js:develop' into levy/median

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/levy/median/src/main.c:
```diff
@@ -0,0 +1,46 @@
+* // returns ~0.795
+*/
+double stdlib_base_dists_levy_median( const double mu, const double c ) {
+    double DENOM = 2.0 * stdlib_base_pow( stdlib_base_erfcinv( 0.5 ), 2.0 );
```
@vivekmaurya001 We should pre-compute this constant as a `static const` in the outer scope since `erfcinv(0.5)` is constant and doesn't change with the inputs.

vivekmaurya001 replied:
ook

vivekmaurya001 replied:
i have tried it with static const but i was getting error in that so i have maded it static double , is it ok ?

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/levy/median/src/main.c:
```diff
@@ -0,0 +1,46 @@
+#include "stdlib/stats/base/dists/levy/median.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/special/erfcinv.h"
+#include "stdlib/math/base/special/pow.h"
+
+
+/**
```
```suggestion
#include "stdlib/math/base/special/pow.h"

/**
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/levy/median/README.md:
```diff
@@ -131,6 +131,101 @@ for ( i = 0; i < 10; i++ ) {
+
+#### stdlib_base_dists_levy_median( mu, c )
+
+Evaluates the [median][median] for a [Lévy][levy-distribution] distribution with location parameter `mu` and scale parameter `c`.
```
```suggestion
Returns the [median][median] for a [Lévy][levy-distribution] distribution with location parameter `mu` and scale parameter `c`.
```

vivekmaurya001 replied:
done

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/levy/median/README.md:
```diff
@@ -131,6 +131,101 @@ for ( i = 0; i < 10; i++ ) {
+    int i;
+
+    for ( i = 0; i < 25; i++ ) {
+        mu = random_uniform( 0.0, 10.0 ) - 5.0;
```
```suggestion
        mu = random_uniform( -5.0, 5.0 );
```

vivekmaurya001 replied:
done

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/levy/median/examples/c/example.c:
```diff
@@ -0,0 +1,40 @@
+	int i;
+
+	for ( i = 0; i < 25; i++ ) {
+		mu = random_uniform( 0.0, 10.0 ) - 5.0;
```
```suggestion
		mu = random_uniform( -5.0, 5.0 );
```

vivekmaurya001 replied:
done

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/levy/median/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,141 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		mu[ i ] = random_uniform( 0.0, 100.0 ) - 50.0;
+		c[ i ] = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		mu[ i ] = random_uniform( -50.0, 50.0 );
		c[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
```

vivekmaurya001 replied:
done , one thing to ask ,can we not take c as 0 or smaller than eps

Planeshifter reviewed (changes requested):
Left a comment on how to optimize the source implementation and pointed out a bunch of other small issues. Should be ready to land once these are addressed.

vivekmaurya001 committed:
Merge branch 'levy/median' of https://github.com/vivekmaurya001/stdlib into levy/median

vivekmaurya001 committed:
Merge branch 'stdlib-js:develop' into levy/median

vivekmaurya001 committed:
Merge branch 'levy/median' of https://github.com/vivekmaurya001/stdlib into levy/median

vivekmaurya001 committed:
feat(add c implementation): add stats/base/dists/levy/median

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: failed
---

vivekmaurya001 committed:
feat(add c implementation): add stats/base/dists/levy/median

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: failed
---

vivekmaurya001 committed:
feat(add c implementation): add stats/base/dists/levy/median

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: failed
---

vivekmaurya001 committed:
feat(add c implementation): add stats/base/dists/levy/median

vivekmaurya001 committed:
feat(add c implementation): add stats/base/dists/levy/median

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: failed
---

vivekmaurya001 committed:
feat(add c implementation): add stats/base/dists/levy/median

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

vivekmaurya001 commented:
@Planeshifter , done the changes please take a look

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/levy/median/src/main.c:
```diff
@@ -0,0 +1,47 @@
+#include "stdlib/math/base/special/erfcinv.h"
+#include "stdlib/math/base/special/pow.h"
+
+static double DENOM ;
```
@vivekmaurya001 This should become `static const double` and we want to pre-compulte the value of `DENOM = 2.0 * stdlib_base_pow( stdlib_base_erfcinv( 0.5 ), 2.0 );` and assign it here. We can evaluate the expression e.g. via the REPL and then hard-code the constant:

```suggestion
static const double DENOM = 0.4549364231195728; // 2.0 * stdlib_base_pow( stdlib_base_erfcinv( 0.5 ), 2.0 )
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/levy/median/src/main.c:
```diff
@@ -0,0 +1,47 @@
+* double y = stdlib_base_levy_median( 9.0, 0.0, 10.0 );
+* // returns ~0.795
+*/
+double stdlib_base_dists_levy_median( const double mu, const double c ) {
+	DENOM = 2.0 * stdlib_base_pow( stdlib_base_erfcinv( 0.5 ), 2.0 );
+	if (
```
```suggestion
double stdlib_base_dists_levy_median( const double mu, const double c ) {
	if (
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/levy/median/src/main.c:
```diff
@@ -0,0 +1,47 @@
+* @return      median
+*
+* @example
+* double y = stdlib_base_levy_median( 9.0, 0.0, 10.0 );
```
```suggestion
* double y = stdlib_base_dists_levy_median( 9.0, 0.0, 10.0 );
```

Planeshifter committed:
refactor: hard-code constant instead of dynamically evaluating it

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
Edited the source implementation in `main.c`, but otherwise PR looked great. Will merge shortly.
